### PR TITLE
Handle dtype/casting/out properly in contractions

### PR DESCRIPTION
**Pull request by manopapad**

### CODE REVIEW

File: cunumeric/array.py [hunk 1]
```diff
@@ -2301,6 +2301,7 @@ def dot(self, rhs, out=None):
             self,
             rhs,
             out=out,
+            casting="no",
```
Comment (magnatelee) [inline cunumeric/array.py:2304]: I assume this is consistent with NumPy's behavior, right?

File: cunumeric/module.py [hunk 2]
```diff
@@ -2658,7 +2658,15 @@ def inner(a, b, out=None):
...
     (a_modes, b_modes, out_modes) = inner_modes(a.ndim, b.ndim)
-    return _contract(a_modes, b_modes, out_modes, a, b, out=out)
+    return _contract(
+        a_modes,
+        b_modes,
+        out_modes,
+        a,
+        b,
+        out=out,
+        casting="unsafe",
```
Comment (magnatelee) [inline cunumeric/module.py:2668]: I'm surprised that the operations use different default casting policies. how did you find out these were the right defaults?

**Review by magnatelee — APPROVED** [review]

File: cunumeric/array.py (hunk 1 above)
Comment (manopapad) [reply cunumeric/array.py:2304]: Yes

**Comment by magnatelee** [comment]
LGTM. feel free to merge.

**Pull request merged**